File: app/src/main/java/it/liceoarzignano/bold/news/NewsListActivity.java
package it.liceoarzignano.bold.news;

import android.app.SearchManager;
import android.content.ComponentName;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.customtabs.CustomTabsClient;
import android.support.customtabs.CustomTabsIntent;
import android.support.customtabs.CustomTabsServiceConnection;
import android.support.customtabs.CustomTabsSession;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.firebase.analytics.FirebaseAnalytics;

import java.util.List;

import it.liceoarzignano.bold.R;
import it.liceoarzignano.bold.firebase.BoldAnalytics;
import it.liceoarzignano.bold.ui.ActionsDialog;
import it.liceoarzignano.bold.ui.recyclerview.RecyclerViewExt;

public class NewsListActivity extends AppCompatActivity {
    private CoordinatorLayout mCoordinator;
    private LinearLayout mEmptyLayout;
    private TextView mEmptyText;

    private NewsHandler mNewsHandler;
    private NewsAdapter mAdapter;
    private CustomTabsClient mClient;
    private CustomTabsSession mCustomTabsSession;
    private CustomTabsServiceConnection mCustomTabsServiceConnection;
    private CustomTabsIntent mCustomTabIntent = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_news_list);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.setNavigationIcon(R.drawable.ic_toolbar_back);
        toolbar.setNavigationOnClickListener(v -> finish());

        mCoordinator = findViewById(R.id.coordinator_layout);
        RecyclerViewExt newsList = findViewById(R.id.news_list);
        mEmptyLayout = findViewById(R.id.news_empty_layout);
        mEmptyText = findViewById(R.id.news_empty_text);

        Intent callingIntent = getIntent();
        long id = callingIntent.getLongExtra("newsId", -1);

        if (id > 0) {
            News2 called = NewsHandler.getInstance(this).get(id);
            if (called != null) {
                showUrl(called.getUrl());
            }
        }

        mNewsHandler = NewsHandler.getInstance(this);

        mAdapter = new NewsAdapter(mNewsHandler.getAll(), this);
        newsList.setAdapter(mAdapter);
    }

    @Override
    protected void onResume() {
        super.onResume();

        // Chrome custom tabs
        setupCCustomTabs();

        String query = null;
        Intent callingIntent = getIntent();
        if (Intent.ACTION_SEARCH.equals(callingIntent.getAction())) {
            query = callingIntent.getStringExtra(SearchManager.QUERY);
        }

        refresh(query);
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (mCustomTabsServiceConnection != null) {
            unbindService(mCustomTabsServiceConnection);
            mCustomTabsServiceConnection = null;
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.search, menu);
        setupSearchView(menu.findItem(R.id.menu_search));
        return true;
    }

    /**
     * Initialize search view
     *
     * @param item menu item
     */
    private void setupSearchView(MenuItem item) {
        SearchView searchView = (SearchView) item.getActionView();

        if (searchView == null) {
            return;
        }

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String mQuery) {
                refresh(mQuery);
                return true;
            }

            @Override
            public boolean onQueryTextChange(String mNewText) {
                refresh(mNewText);
                return true;
            }
        });
    }

    /**
     * Refresh list
     *
     * @param query search query
     */
    private void refresh(String query) {
        final List<News2> news = mNewsHandler.getByQuery(query);
        mAdapter.updateList(news);
        mEmptyLayout.setVisibility(news.isEmpty() ? View.VISIBLE : View.GONE);
        mEmptyText.setText(getString(query != null && !query.isEmpty() ?
                R.string.search_no_result : R.string.news_empty));

    }

    /**
     * Initialize chrome custom tabs
     */
    private void setupCCustomTabs() {
        if (mCustomTabIntent != null) {
            return;
        }

        mCustomTabsServiceConnection = new CustomTabsServiceConnection() {
            @Override
            public void onCustomTabsServiceConnected(ComponentName componentName,
                                                     CustomTabsClient customTabsClient) {
                mClient = customTabsClient;
                mClient.warmup(0L);
                mCustomTabsSession = mClient.newSession(null);
            }

            @Override
            public void onServiceDisconnected(ComponentName name) {
                        mClient = null;
                    }
        };

        CustomTabsClient.bindCustomTabsService(this, "com.android.chrome",
                mCustomTabsServiceConnection);

        mCustomTabIntent = new CustomTabsIntent.Builder(mCustomTabsSession)
                .setToolbarColor(ContextCompat.getColor(this, R.color.colorPrimary))
                .setShowTitle(true)
                .setStartAnimations(this, R.anim.slide_in_right, R.anim.slide_out_left)
                .setExitAnimations(this, android.R.anim.slide_in_left,
                        android.R.anim.slide_out_right)
                .build();
    }

    /**
     * Open the url associated with a news
     *
     * @param url website url
     */
    void showUrl(String url) {
        new BoldAnalytics(this).log(FirebaseAnalytics.Event.VIEW_ITEM, "News url");
        setupCCustomTabs();
        mCustomTabIntent.launchUrl(this, Uri.parse(url));
    }

    @SuppressWarnings("SameReturnValue")
    boolean newsActions(News2 news) {
        ActionsDialog dialog = new ActionsDialog(this, false, news.getId());
        dialog.setOnActionsListener(new ActionsDialog.OnActionsDialogListener() {
            @Override
            public void onShare() {
                String message = String.format("%1$s (%2$s)\n%3$s", news.getTitle(),
                        news.getDate(), news.getDescription());
                startActivity(Intent.createChooser(new Intent(Intent.ACTION_SEND)
                                .setType("text/plain")
                                .putExtra(Intent.EXTRA_TEXT, message),
                        getString(R.string.share_title)));
            }

            @Override
            public void onDelete() {
                mNewsHandler.delete(news.getId());
                Snackbar.make(mCoordinator, getString(R.string.actions_removed), Snackbar.LENGTH_LONG)
                        .show();
                refresh(null);
            }
        });
        dialog.show();

        return true;
    }
}

File: app/src/main/kotlin/it/liceoarzignano/bold/news/NewsListActivity.kt
package it.liceoarzignano.bold.news

import android.app.SearchManager
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.browser.customtabs.CustomTabsClient
import androidx.browser.customtabs.CustomTabsIntent
import androidx.browser.customtabs.CustomTabsServiceConnection
import androidx.browser.customtabs.CustomTabsSession
import androidx.core.content.ContextCompat
import com.google.android.material.snackbar.Snackbar
import it.liceoarzignano.bold.R
import it.liceoarzignano.bold.ui.ActionsDialog
import it.liceoarzignano.bold.ui.recyclerview.RecyclerViewExt
import it.liceoarzignano.bold.ui.recyclerview.ScrollListener
import it.liceoarzignano.bold.utils.SystemUtils
import it.liceoarzignano.bold.utils.Time

class NewsListActivity : AppCompatActivity() {
    private lateinit var mCoordinator: androidx.coordinatorlayout.widget.CoordinatorLayout
    private lateinit var mEmptyLayout: LinearLayout
    private lateinit var mEmptyText: TextView

    private lateinit var mNewsHandler: NewsHandler
    private lateinit var mAdapter: NewsAdapter
    private var mClient: CustomTabsClient? = null
    private var mCustomTabsSession: CustomTabsSession? = null
    private var mCustomTabsServiceConnection: CustomTabsServiceConnection? = null
    private var mCustomTabIntent: CustomTabsIntent? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_news_list)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        toolbar.setNavigationIcon(R.drawable.ic_toolbar_back)
        toolbar.setNavigationOnClickListener { finish() }

        mCoordinator = findViewById(R.id.coordinator_layout)
        val newsList = findViewById<RecyclerViewExt>(R.id.news_list)
        mEmptyLayout = findViewById(R.id.news_empty_layout)
        mEmptyText = findViewById(R.id.news_empty_text)

        val callingIntent = intent
        val url = callingIntent.getStringExtra("newsUrl")

        if (url != null && !url.isEmpty()) {
            showUrl(url)
        }

        mNewsHandler = NewsHandler.getInstance(baseContext)
        mAdapter = NewsAdapter(mNewsHandler.all, this)
        newsList.adapter = mAdapter
        if (SystemUtils.isNotLegacy) {
            newsList.addOnScrollListener(ScrollListener(toolbar, resources, newsList.top))
        }
    }

    override fun onResume() {
        super.onResume()

        // Chrome custom tabs
        setupCCustomTabs()

        var query: String? = null
        val callingIntent = intent
        if (Intent.ACTION_SEARCH == callingIntent.action) {
            query = callingIntent.getStringExtra(SearchManager.QUERY)
        }

        refresh(query)
    }

    override fun onPause() {
        super.onPause()
        if (mCustomTabsServiceConnection != null) {
            unbindService(mCustomTabsServiceConnection as ServiceConnection)
            mCustomTabsServiceConnection = null
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.search, menu)
        setupSearchView(menu.findItem(R.id.menu_search))
        return true
    }

    private fun setupSearchView(item: MenuItem) {
        val searchView = item.actionView as SearchView

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(mQuery: String): Boolean {
                refresh(mQuery)
                return true
            }

            override fun onQueryTextChange(mNewText: String): Boolean {
                refresh(mNewText)
                return true
            }
        })
    }

    private fun refresh(query: String?) {
        val news = mNewsHandler.getByQuery(query?.toLowerCase())
        mAdapter.updateList(news)
        mEmptyLayout.visibility = if (news.isEmpty()) View.VISIBLE else View.GONE
        mEmptyText.text = getString(if (query != null && !query.isEmpty())
            R.string.search_no_result
        else
            R.string.news_empty)

    }

    private fun setupCCustomTabs() {
        if (mCustomTabIntent != null) {
            return
        }

        mCustomTabsServiceConnection = object : CustomTabsServiceConnection() {
            override fun onCustomTabsServiceConnected(componentName: ComponentName,
                                                      customTabsClient: CustomTabsClient) {
                mClient = customTabsClient
                mClient!!.warmup(0L)
                mCustomTabsSession = mClient!!.newSession(null)
            }

            override fun onServiceDisconnected(name: ComponentName) {
                mClient = null
            }
        }

        CustomTabsClient.bindCustomTabsService(this, "com.android.chrome",
                mCustomTabsServiceConnection)

        mCustomTabIntent = CustomTabsIntent.Builder(mCustomTabsSession)
                .setToolbarColor(ContextCompat.getColor(this, R.color.colorPrimary))
                .setShowTitle(true)
                .setStartAnimations(this, R.anim.slide_in_right, R.anim.slide_out_left)
                .setExitAnimations(this, android.R.anim.slide_in_left,
                        android.R.anim.slide_out_right)
                .build()
    }

    internal fun showUrl(url: String) {
        setupCCustomTabs()
        mCustomTabIntent!!.launchUrl(this, Uri.parse(url))
    }

    internal fun newsActions(news: News): Boolean {
        val dialog = ActionsDialog(this, false, false, news.id)
        dialog.setOnActionsListener(object : ActionsDialog.OnActionsDialogListener {
            override fun onShare() {
                val message = String.format("%1\$s (%2\$s)\n%3\$s", news.title,
                        Time(news.date), news.description)
                startActivity(Intent.createChooser(Intent(Intent.ACTION_SEND)
                        .setType("text/plain")
                        .putExtra(Intent.EXTRA_TEXT, message),
                        getString(R.string.share_title)))
            }

            override fun onDelete() {
                mNewsHandler.delete(news.id)
                Snackbar.make(mCoordinator, getString(R.string.actions_removed), Snackbar.LENGTH_LONG)
                        .show()
                refresh(null)
            }
        })
        dialog.show()

        return true
    }
}
